iroha_introspect/src/lib.rs

```diff
@@ -0,0 +1,358 @@
+//! Module for introspecting rust types in other languages for translation.
+
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    time::Duration,
+};
+
+/// Derive introspect. It will make your structure introspectable
+pub use iroha_introspect_derive::Introspect;
+use serde::Serialize;
+
+/// Metadata map
+pub type MetaMap = BTreeMap<String, Metadata>;
+
+/// Introspect trait
+pub trait Introspect {
+    /// Returns unique type name.
+    /// WARN: `std::any::type_name` is compiler related, so is not unique.
+    /// I guess we should change it somehow later
+    fn type_name() -> String {
+        let mut name = module_path!().to_owned();
+        name.push_str("::");
+        name.push_str(std::any::type_name::<Self>());
+        name
+    }
+
+    /// Returns info about current type. Will return map from type names to its metadata
+    fn get_introspect() -> MetaMap {
+        let mut map = MetaMap::new();
+        Self::introspect(&mut map);
+        map
+    }
+
+    /// Introspect function. Give it empty map and it will return description of types
+    /// related to this type
+    fn introspect(metamap: &mut MetaMap);
+}
+
+/// Metadata
+#[derive(Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Serialize, Clone)]
+pub enum Metadata {
+    /// Named structure
+    NamedStruct(NamedFieldsMeta),
```

Comment:
With what name such meta will be serialized into json? As "NamedStruct"?

Reply:
it will look like this:
```json
  "iroha_crypto::PublicKey": {
    "NamedStruct": {
      "declarations": [
        {
          "name": "digest_function",
          "ty": "String"
        },
        {
          "name": "payload",
          "ty": "Vec<u8>"
        }
      ]
    }
  },
```

Reply:
"NamedStruct" and "UnnamedStruct" is a bad name. Let's rename it as discussed earlier.

Reply:
Guess we should do something with naming. I think we should still preserve word struct in name. Beforehand it was about fields themselves (they got renamed though).

Idk how to name it to show that it is StructWithNamedFields

Reply:
Yeah, that's much better. Or it can be just "NamedFields" or "UnnamedFields". Up to you

Reply:
I mean it sounds java-like. It was just a semantic caricature for this. Idk how to preserve most of properties via another naming.

Reply:
rkharisov> We can name them CLikeStruct and TupleStruct
i1i1> Sounds better, but CLikeStruct gives wrong semantical context. TupleStruct is fine.